### DISCUSSION TITLE
Suggestion: Use a GDB remotely for debugging

### DISCUSSION THREAD

**Discussion by GravisZro** [itsmattkc/dotnet9x#24] [Ideas]
While you may be quite limited by what debuggers will run on Windows 9x, you can bypass this problem by using a [GDB stub](https://sourceware.org/gdb/current/onlinedocs/gdb.html/Remote-Debugging.html#Remote-Debugging) (like [this one](https://github.com/mborgerson/gdbstub)) and debugging remotely from the VM host OS. A remote GDB session means you can use any GDB capable debugger including [VS Code](https://stackoverflow.com/questions/38089178/is-it-possible-to-attach-to-a-remote-gdb-target-with-vscode#answers) and probably Ghidra.

I forgot (because I haven't done remote win32 debugging in a long time) but there are [prebuilt mingw64 32-bit builds of GDB](https://sourceforge.net/projects/mingw-w64/files/External%20binary%20packages%20(Win64%20hosted)/gdb/) and the really old [mingw builds of GDB](https://sourceforge.net/projects/mingw/files/MinGW/Extension/gdb/)(but they need mingw DLLs IIRC) which are suitable for the same purpose.

I hope this suggestion helps.

**Comment by friendlyanon** [comment]
Building MinGW is trivial with [w64devkit](https://github.com/skeeto/w64devkit) (which is also [my preferred MinGW distribution](https://github.com/friendlyanon/w64devkit)) and it produces statically linked binaries. It also includes a patch if you want to target i686. I'm not sure that or the prebuilt you linked are okay for this purpose, since Windows 95 requires at least a 386. Modifying the dockerfile and patch of w64devkit would be a lot easier than fishing for the right prebuilt distribution.

**Comment by itsmattkc** [comment]
I had considered doing this but wasn't confident of whether I'd be able to find any compatible GDB build that supported 95 (I'm not even sure if I had managed to get the network working in my VM). But it may be an avenue worth exploring for future work.
